### ISSUE TITLE
ipfs add concurrency

### ISSUE THREAD

**Issue by crackcomm** [ipfs/go-ipfs#4080]
Go natural concurrency model should let do it easily.

**Comment by Stebalien** [comment]
Is `ipfs add` not IO bound?

**Comment by crackcomm** [comment]
Even though it is, it would speed up things a little (sometimes).

**Comment by Kubuxu** [comment]
It is very IO bound, I know @whyrusleeping had some concurrent hashing patches for testing but the increase wasn't that great.

**Comment by Stebalien** [comment]
Memory mapping files instead of reading them and prefetching to max out our IO bandwidth may help. Unfortunately, doing this is **hard**.

**Comment by Kubuxu** [comment]
Yes, I was thinking about it for a long time, we can do that on Linux/Unix by creating Unix socket API and doing file descriptor send over it, then we can memory map that file.
